src/apps/relay/ns_ioalib_engine_impl.c

```diff
@@ -3106,105 +3106,114 @@ int send_data_from_ioa_socket_nbh(ioa_socket_handle s, ioa_addr *dest_addr, ioa_
-                                                */
-                s->in_write = 0;
-              } else {
-                // drop the packet
-                ;
-              }
-            }
-          } else if (s->ssl) {
-            send_ssl_backlog_buffers(s);
-            ret = ssl_send(s, (char *)ioa_network_buffer_data(nbh), ioa_network_buffer_get_size(nbh),
-                           (s->e ? s->e->verbose : TURN_VERBOSE_NONE));
-            if (ret < 0)
-              s->tobeclosed = 1;
-            else if (ret == 0)
-              add_buffer_to_buffer_list(&(s->bufs), (char *)ioa_network_buffer_data(nbh),
-                                        ioa_network_buffer_get_size(nbh));
-          } else if (s->fd >= 0) {
-
-            if (s->connected && !(s->parent_s)) {
-              dest_addr = NULL; /* ignore dest_addr */
-            } else if (!dest_addr) {
-              dest_addr = &(s->remote_addr);
-            }
+      if (!tcp_congestion_control || is_socket_writeable(s, (size_t)ret, __FUNCTION__, 2)) {
+        s->in_write = 1;
+        if (bufferevent_write(s->bev, ioa_network_buffer_data(nbh), ioa_network_buffer_get_size(nbh)) < 0) {
+          ret = -1;
+          perror("bufev send");
+          log_socket_event(s, "socket write failed, to be closed", 1);
+          s->tobeclosed = 1;
+          s->broken = 1;
+        }
+        /*
+        bufferevent_flush(s->bev,
+                                        EV_READ|EV_WRITE,
+                                        BEV_FLUSH);
+                                        */
+        s->in_write = 0;
+      } else {
+        // drop the packet
+        ;
+      }
 
-            ret = udp_send(s, dest_addr, (char *)ioa_network_buffer_data(nbh), ioa_network_buffer_get_size(nbh));
-            if (ret < 0) {
-              s->tobeclosed = 1;
-#if defined(EADDRNOTAVAIL)
-              int perr = socket_errno();
-#endif
-              perror("udp send");
-#if defined(EADDRNOTAVAIL)
-              if (dest_addr && (perr == EADDRNOTAVAIL)) {
-                char sfrom[129];
-                addr_to_string(&(s->local_addr), (uint8_t *)sfrom);
-                char sto[129];
-                addr_to_string(dest_addr, (uint8_t *)sto);
-                TURN_LOG_FUNC(TURN_LOG_LEVEL_ERROR, "%s: network error: address unreachable from %s to %s\n",
-                              __FUNCTION__, sfrom, sto);
-              }
-#endif
-            }
-          }
+      break;
+    }
+
+    if (s->ssl) {
+      send_ssl_backlog_buffers(s);
+      ret = ssl_send(s, (char *)ioa_network_buffer_data(nbh), ioa_network_buffer_get_size(nbh),
+                     (s->e ? s->e->verbose : TURN_VERBOSE_NONE));
+      if (ret < 0)
+        s->tobeclosed = 1;
+      else if (ret == 0)
+        add_buffer_to_buffer_list(&(s->bufs), (char *)ioa_network_buffer_data(nbh), ioa_network_buffer_get_size(nbh));
+
+      break;
+    }
+
+    if (s->fd >= 0) {
```

Comment:
if a file descriptor of 0 is valid, what does a negative file descriptor represent?